atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/RetriableManyHostPoolingContainer.java

```diff
@@ -103,39 +107,49 @@ public static ManyHostPoolingContainer create(int maxTries, int triesBeforeRando
-                || e instanceof TTransportException
-                || e instanceof TimedOutException
-                || e instanceof SocketTimeoutException
-                || e instanceof UnavailableException
-                || e instanceof InsufficientConsistencyException) {
+        if (isRetriableException(e)) {
             if (numTries >= maxTries) {
                 if (e instanceof TTransportException
                         && e.getCause() != null
                         && (e.getCause().getClass() == SocketException.class)) {
-                    String msg = "Error writing to Cassandra socket. Likely cause: Exceeded maximum thrift frame size; unlikely cause: network issues.";
-                    log.error("Tried to connect to cassandra " + numTries + " times. " + msg, e);
+                    String msg = "Error writing to Cassandra socket. Likely cause: " +
+                            "Exceeded maximum thrift frame size; unlikely cause: network issues.";
+                    log.error("Tried to connect to Cassandra {} times. {}", numTries, msg, e);
                     e = new TTransportException(((TTransportException) e).getType(), msg, e);
                 } else {
-                    log.error("Tried to connect to cassandra " + numTries + " times.", e);
+                    log.error("Tried to connect to Cassandra {} times.", numTries, e);
                 }
                 throw (K) e;
             } else {
-                log.warn("Error occurred talking to cassandra. Attempt {} of {}.", numTries, maxTries, e);
-                if (e instanceof SocketTimeoutException
-                        || e instanceof UnavailableException) {
+                log.warn("Error occurred talking to Cassandra. Attempt {} of {}.", numTries, maxTries, e);
+                if (isConnectionException(e)) {
                     // Connection is no good? This may be due to a long GC, we should back off sending them requests.
                     // Binary exponential backoff; should in total take ~10s on average w/maxTries=10
                     try {
-                        Thread.sleep(new Random().nextInt((1 << numTries)-1)*20);
-                    } catch (InterruptedException e1) {
+                        Thread.sleep(ThreadLocalRandom.current().nextInt((1 << numTries) - 1) * 20);
+                    } catch (InterruptedException ignored) {
                         Thread.currentThread().interrupt();
                     }
                 }
             }
         } else {
+            log.error("Unable to connect to Cassandra: " + e);
             throw (K) e;
         }
     }
 
+    private static boolean isConnectionException(Exception e) {
+        return e instanceof SocketTimeoutException
+                || e instanceof UnavailableException
+                || e instanceof NoSuchElementException;
+    }
+
+    private static boolean isRetriableException(Exception e) {
+        return isConnectionException(e)
```

Comment:
Because we should be retrying and backing off for `NoSuchElementException` as that occurs when there is no connection currently available in the Cassandra client pool.